### DISCUSSION TITLE
A few questions regarding how to use this API

### DISCUSSION THREAD

**Discussion by xiaocheng0305** [weather-gov/api#229] [Q&A]
Hi,

I have a few questions regarding how to use this API.

1. If I sent a request from my code like this:

Response = requests.get("https://api.weather.gov/points/%f,%f" % (29.3,-94.9))

I'm getting SSLCertVerificationError.
I assume we need an API key to use this service, but I haven't found anyway to create an account on the weather.gov website or  how to get an API key. 
I tried to add a header and include User-Agent in the header but that didn't help either.
e.g.

response = requests.get(
	'https://api.weather.gov/points/30,-90',
	headers={'Accept': 'application/vnd.noaa.dwml+xml', 'User-Agent': 'CERN-LineMode/2.15 libwww/2.17b3'}
)

So what am I missing?
To work around this temporarily, I turned off certificate verification:
weatherResponse: Response = requests.get(
        "https://api.weather.gov/points/%f,%f" % (29.3,-94.9), verify=False)

That allows me to access the API, but of course it's not ideal. And maybe this causes another problem which I will mention below.

2. This API seems to be designed to provide Pasquill stability class, but doesn't actually provide the value. i.e. If I visit here:
https://api.weather.gov/gridpoints/HGX/81,79
At the bottom of the page there is:

"stability": {
            "values": []
        },

Is it possible to get the Pasquill stability class from the API?

3. As mentioned above, I'm able to access the API if I turn of certificate verification. But I'm only able to access up to /forecast/hourly
e.g.
https://api.weather.gov/gridpoints/HGX/81,79/forecast/hourly

This doesn't provide the Pasquill stability class, nor solar radiation / cloud coverage which can be used to calculate the stability class. So I need to access e.g.:
https://api.weather.gov/gridpoints/HGX/81,79/
which has a section for "skyCover". I assume that's equivalent to cloud coverage.

However when I try to access this site in my code, I'm not getting any meaningful values back for skyCover or any other variables:
![GitHub question](https://user-images.githubusercontent.com/79928559/109683148-fb746a00-7b76-11eb-83de-94bce50756d8.png)

Not sure whether it's because I turned off certificate verification. I am able to access /forecast and /forecast/hourly though.

Much appreciated for your help!

best regards

**Comment by jfbauer432** [comment]
It looks like you are using python and the requests module.  That is what I am using and have not run into the SSL issue you mentioned.  Nor have I included an API key (though I hear that is coming eventually).  I tried just what you did and it worked for me (python 3.8.6).


```
>>> import requests
>>> Response = requests.get("https://api.weather.gov/points/%f,%f" % (29.3,-94.9))
>>> Response
<Response [200]>
>>> Response.headers
{'Server': 'nginx/1.16.1', 'Content-Type': 'application/geo+json', 'Access-Control-Allow-Origin': '*', 'Access-Control-Allow-Headers': 'Feature-Flags', 'X-Request-ID': '35d7143a-99d3-4cc3-bd8e-4c73a85f076e', 'X-Correlation-ID': '44c5462b', 'X-Server-ID': 'vm-bldr-nids-apiapp4.ncep.noaa.gov', 'Content-Encoding': 'gzip', 'Content-Length': '774', 'Cache-Control': 'public, max-age=86069, s-maxage=120', 'Expires': 'Wed, 03 Mar 2021 18:51:50 GMT', 'Date': 'Tue, 02 Mar 2021 18:57:21 GMT', 'Connection': 'keep-alive', 'Vary': 'Accept-Encoding, Accept,Feature-Flags,Accept-Language', 'X-Edge-Request-ID': '1c854c36', 'Strict-Transport-Security': 'max-age=31536000 ; includeSubDomains ; preload'}
>>> Response.json()
{'@context': ['https://geojson.org/geojson-ld/geojson-context.jsonld', {'@version': '1.1', 'wx': 'https://api.weather.gov/ontology#', 's': 'https://schema.org/', 'geo': 'http://www.opengis.net/ont/geosparql#', 'unit': 'http://codes.wmo.int/common/unit/', '@vocab': 'https://api.weather.gov/ontology#', 'geometry': {'@id': 's:GeoCoordinates', '@type': 'geo:wktLiteral'}, 'city': 's:addressLocality', 'state': 's:addressRegion', 'distance': {'@id': 's:Distance', '@type': 's:QuantitativeValue'}, 'bearing': {'@type': 's:QuantitativeValue'}, 'value': {'@id': 's:value'}, 'unitCode': {'@id': 's:unitCode', '@type': '@id'}, 'forecastOffice': {'@type': '@id'}, 'forecastGridData': {'@type': '@id'}, 'publicZone': {'@type': '@id'}, 'county': {'@type': '@id'}}], 'id': 'https://api.weather.gov/points/29.3,-94.9', 'type': 'Feature', 'geometry': {'type': 'Point', 'coordinates': [-94.9, 29.3]}, 'properties': {'@id': 'https://api.weather.gov/points/29.3,-94.9', '@type': 'wx:Point', 'cwa': 'HGX', 'forecastOffice': 'https://api.weather.gov/offices/HGX', 'gridId': 'HGX', 'gridX': 82, 'gridY': 75, 'forecast': 'https://api.weather.gov/gridpoints/HGX/82,75/forecast', 'forecastHourly': 'https://api.weather.gov/gridpoints/HGX/82,75/forecast/hourly', 'forecastGridData': 'https://api.weather.gov/gridpoints/HGX/82,75', 'observationStations': 'https://api.weather.gov/gridpoints/HGX/82,75/stations', 'relativeLocation': {'type': 'Feature', 'geometry': {'type': 'Point', 'coordinates': [-94.914945, 29.297265]}, 'properties': {'city': 'Tiki Island', 'state': 'TX', 'distance': {'value': 1480.7988614439, 'unitCode': 'unit:m'}, 'bearing': {'value': 78, 'unitCode': 'unit:degrees_true'}}}, 'forecastZone': 'https://api.weather.gov/zones/forecast/GMZ335', 'timeZone': 'America/Chicago', 'radarStation': 'KHGX'}}
>>> requests.__version__
'2.23.0'
```

I also looked at the skyCover, and got what look like decent results

```
>>> r = requests.get('https://api.weather.gov/gridpoints/HGX/81,79')
>>> vals = r.json()['properties']['skyCover']['values']
>>> for x in vals[:10]:
...     print(x['value'])
... 
95
84
87
90
72
53
28
0
1
3

```

Are you still having issues?  If not, perhaps there was some sort of outage earlier.

**Comment by xiaocheng0305** [comment]
Thanks for your help @jfbauer432 !

Yes I'm still having the issues, I've been having the issues for a few days. I'm using python 3.9 64bit in PyCharm. Probably I should try another IDE if the code is working for you?

The errors I got, more specifically, looks like this:
![image](https://user-images.githubusercontent.com/79928559/109704870-04713580-7b8f-11eb-8e49-44a2bdf58923.png)

It's thrown when running this line:
Response = requests.get("https://api.weather.gov/points/%f,%f" % (29.3,-94.9))

**Comment by jfbauer432** [comment]
I don't know anything about PyCharm, but I verified that python 3.9.2 also works for me.

I am suspecting you might be picking up old or bad certificates.  Looks like requests, uses module certifi for the certs.  You can import that and print certifi.__version__ and see what you get.  For comparison, on Ubuntu 20.10, I have 2020.04.05.1.  In a python 3.9.2 docker image I get 2020.12.05.  On a raspberry pi system with python 3.7.2, certifi.__version__ is 2018.08.24.  In all of those, the example I tried all worked.  You can also print out certifi.where() to see where it is looking for the certs and if that file even exists on your system.

If you haven't already, try a simple example to be sure nothing in you code is adversely affecting how requests is working.

**Comment by xiaocheng0305** [comment]
Thanks again @jfbauer432 !
I am able to request from other APIs. My certifi version is 2020.12.05 which you've got on your docker image. The file does exist at the location that certifi.where() returns... A bit stuck now.

**Comment by scadergit** [comment]
The API does not require a key (yet), only a user agent header. The sky cover invalid values were due to a temporary issue with MADIS that feeds the API oberservations.

I'm not familiar enough with PyCharm to know if validation is something special it's attempting to do, but I'll ask the team tomorrow that are more familiar with Python.
